File: refactor_version/RayTracing/cpp_impl/src/objects.cpp
#include "constants.h"
#include "objects.h"

#include <cmath>
#include <glm/glm.hpp>
#include <glm/gtc/constants.hpp>
#include <glm/gtx/norm.hpp>

/* ========== ========== ========== */
/* Sphere */
/* ========== ========== ========== */

Sphere::Sphere(
    const glm::vec3& position,
    const glm::vec3& color,
    const float reflection, 
    const float diffuse,
    const float specular_coef,
    const float specular_k,
    const float radius
): BaseObject(position, color, reflection, diffuse, specular_coef, specular_k), radius(radius) {}

float Sphere::intersect(const glm::vec3& origin, const glm::vec3& direction) const {
    glm::vec3 OC = position - origin;

    if (glm::length(OC) < radius || glm::dot(OC, direction) < 0){
        return INFINITY;
    }

    float l = std::abs(glm::dot(OC, direction));

    float m_square = glm::length(OC)*glm::length(OC) - l*l;
    float q_square = radius*radius - m_square;

    float dist = 0<=q_square ? l-std::sqrt(q_square) : INFINITY;
    return dist;
}

glm::vec3 Sphere::get_normal(const glm::vec3& point) const {
    return glm::normalize(point - position);
}

/* ========== ========== ========== */
/* Plane */
/* ========== ========== ========== */

Plane::Plane(
    const glm::vec3& position,
    const glm::vec3& color,
    float reflection,
    float diffuse,
    float specular_coef,
    float specular_k,
    const glm::vec3& normal
): BaseObject(position, color, reflection, diffuse, specular_coef, specular_k), normal(glm::normalize(normal)) {}

float Plane::intersect(const glm::vec3& origin, const glm::vec3& direction) const {
    float dn = glm::dot(direction, normal);
    if (std::abs(dn) < EPSILON) {
        return INFINITY;
    }

    float d = glm::dot(position-origin, normal) / dn;

    float dist = 0 < d ? d : INFINITY;

    return dist;
}

glm::vec3 Plane::get_normal(const glm::vec3& /*points*/) const {
    return normal;
}

File: refactor_version/RayTracing/py_impl/objects.py
from .constants import *
from .fns import *
from .base_object import BaseObject
import numpy as np

class Sphere(BaseObject):
    def __init__(
            self,
            position: list[float], # center of the sphere
            color: list[float],
            reflection: float,
            diffuse: float,
            specular_coef: float,
            specular_k: float,
            radius: float,
        ):
        super().__init__(position, color, reflection, diffuse, specular_coef, specular_k)
        self.radius = radius

    def intersect(self, origin: np.ndarray, direction: np.ndarray):

        OC = self.position - origin
        
        if (norm(OC) < self.radius) or (np.dot(OC, direction) < 0):
            return INF
        
        l = np.abs(np.dot(OC, direction))

        m_square = norm(OC)**2 - l**2
        q_square = self.radius**2 - m_square
        
        return (l - np.sqrt(q_square)) if q_square >= 0 else INF

    def get_normal(self, point):
        return normalize(point - self.position)

class Plane(BaseObject):
    def __init__(
            self,
            position: list[float],
            color: list[float],
            reflection: float,
            diffuse: float,
            specular_coef: float,
            specular_k: float,
            normal: list[float],
        ):
        super().__init__(position, color, reflection, diffuse, specular_coef, specular_k)
        self.normal = normalize(np.array(normal))

    def intersect(self, origin: np.ndarray, direction: np.ndarray):

        dn = np.dot(direction, self.normal)       
        if np.abs(dn) < EPSILON:
            return INF
        
        d = np.dot(self.position - origin, self.normal) / dn
        return d if d > 0 else INF

    def get_normal(self, *x):
        return self.normal